File: libs/pilight/hardware/hardware.c
/*
  Copyright (C) CurlyMo

  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <fcntl.h>
#include <errno.h>
#ifndef _WIN32
	#include <unistd.h>
	#include <regex.h>
	#include <sys/ioctl.h>
	#include <dlfcn.h>
	#ifdef __mips__
		#define __USE_UNIX98
	#endif
	#include <pthread.h>
	#include <libgen.h>
	#include <dirent.h>
#endif
#include <sys/stat.h>
#include <time.h>
#include <assert.h>

#include "../core/pilight.h"
#include "../core/common.h"
#include "../core/log.h"
#include "../core/json.h"
#include "../core/dso.h"
#include "../config/settings.h"

static int init = 0;

int hardware_gc(void) {
	init = 0;

	logprintf(LOG_DEBUG, "garbage collected config hardware library");
	return EXIT_SUCCESS;
}

void hardware_init(void) {
	if(init == 1) {
		return;
	}
	init = 1;
	plua_init();

	char path[PATH_MAX];
	char *f = STRDUP(__FILE__);
	struct dirent *file = NULL;
	DIR *d = NULL;
	struct stat s;
	char *hardware_root = HARDWARE_ROOT;

	if(f == NULL) {
		OUT_OF_MEMORY /*LCOV_EXCL_LINE*/
	}

	struct lua_state_t *state = plua_get_free_state();
	int ret = config_setting_get_string(state->L, "hardware-root", 0, &hardware_root);
	assert(plua_check_stack(state->L, 0) == 0);
	plua_clear_state(state);

	if((d = opendir(hardware_root))) {
		while((file = readdir(d)) != NULL) {
			memset(path, '\0', PATH_MAX);
			sprintf(path, "%s%s", hardware_root, file->d_name);
			if(stat(path, &s) == 0) {
				/* Check if file */
				if(S_ISREG(s.st_mode)) {
					if(strstr(file->d_name, ".lua") != NULL) {
						plua_module_load(path, HARDWARE);
					}
				}
			}
		}
	}
	if(ret == 0 || hardware_root != (void *)HARDWARE_ROOT) {
		FREE(hardware_root);
	}
	closedir(d);
	FREE(f);
}

File: libs/pilight/config/json/hardware.lua
-- Copyright (C) 2013 - 2016 CurlyMo

-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at http://mozilla.org/MPL/2.0/.

local json = require "json";
local dump = require "dump";

local M = {}

function firstToUpper(str)
    return (str:gsub("^%l", string.upper))
end

function M.read(f)
	if f == nil then
		return 0;
	end

	local file = pilight.io.file(f);

	file.open("r");
	local content = '';
	for line in file.readline() do
		content = content .. line;
	end
	file.close();

	local config = pilight.config();
	local data = config.getData();
	local jobject = json.parse(content);
	local hardware = jobject['hardware'];

	data['hardware'] = hardware;

	if hardware ~= nil then
		for k, v in pairs(hardware) do
			local obj = config.getHardware(k);
			if obj == nil then
				return 0;
			end
			if obj.validate() == false then
				return 0;
			end
		end;
	end;

	return 1;
end

function M.write()
	return nil;
end

function M.info()
	return {
		name = "hardware",
		version = "1.0",
		reqversion = "5.0",
		reqcommit = "87"
	}
end

return M;
